src/internal/_equals.js

```diff
@@ -1,114 +1,196 @@
+    }
+
+    switch (type(a)) {
+      case 'Arguments':
+      case 'Array':
+      case 'Object':
+        if (typeof a.constructor === 'function' &&
           _functionName(a.constructor) === 'Promise') {
-        return a === b;
-      }
-      break;
-    case 'Boolean':
-    case 'Number':
-    case 'String':
-      if (!(typeof a === typeof b && identical(a.valueOf(), b.valueOf()))) {
-        return false;
-      }
-      break;
-    case 'Date':
-      if (!identical(a.valueOf(), b.valueOf())) {
-        return false;
-      }
-      break;
-    case 'Error':
-      return a.name === b.name && a.message === b.message;
-    case 'RegExp':
-      if (!(a.source === b.source &&
+          return a === b;
+        }
+        break;
+      case 'Boolean':
+      case 'Number':
+      case 'String':
+        if (!(typeof a === typeof b && identical(a.valueOf(), b.valueOf()))) {
+          return false;
+        }
+        break;
+      case 'Date':
+        if (!identical(a.valueOf(), b.valueOf())) {
+          return false;
+        }
+        break;
+      case 'Error':
+        return a.name === b.name && a.message === b.message;
+      case 'RegExp':
+        if (!(a.source === b.source &&
             a.global === b.global &&
             a.ignoreCase === b.ignoreCase &&
             a.multiline === b.multiline &&
             a.sticky === b.sticky &&
             a.unicode === b.unicode)) {
-        return false;
-      }
-      break;
-    case 'Map':
-    case 'Set':
-      if (!_equals(_arrayFromIterator(a.entries()), _arrayFromIterator(b.entries()), stackA, stackB)) {
-        return false;
-      }
-      break;
-    case 'Int8Array':
-    case 'Uint8Array':
-    case 'Uint8ClampedArray':
-    case 'Int16Array':
-    case 'Uint16Array':
-    case 'Int32Array':
-    case 'Uint32Array':
-    case 'Float32Array':
-    case 'Float64Array':
-      break;
-    case 'ArrayBuffer':
-      break;
-    default:
-      // Values of other types are only equal if identical.
-      return false;
-  }
+          return false;
+        }
+        break;
+      case 'Map':
+        if (a.size !== b.size) {
+          return false;
+        }
 
-  var keysA = keys(a);
-  if (keysA.length !== keys(b).length) {
-    return false;
-  }
+        var stackEquals = _stackEquals(a, b, stackA, stackB);
+        if (stackEquals !== undefined) {
+          return stackEquals;
+        }
+
+        stackA.push(a);
+        stackB.push(b);
+
+        // this optimization can be not needed, depends on use case.
+        if (!_uniqContentEquals(a.keys(), b.keys(), stackA, stackB)) {
+          return false;
+        }
+
+        if (!_uniqContentEquals(a.entries(), b.entries(), stackA, stackB)) {
+          return false;
+        }
+
+        stackA.pop();
+        stackB.pop();
+        return true;
+      case 'Set':
+        if (a.size !== b.size) {
+          return false;
+        }
+
+        var idx = stackA.length - 1;
```

Comment:
We don't need this here anymore since you are calling your new `_stackEquals` function below.